From bc5662cfc1abe5c9a53b7295ab0ced523a4f8ef2 Mon Sep 17 00:00:00 2001
From: Michal Hrusecky <michal.hrusecky@nic.cz>
Date: Thu, 13 Oct 2016 10:41:20 +0200
Subject: [PATCH] Fix crash when calling txpowerlist on ath10k set to 2.4GHz

Calling iwinfo phy0 txpowerlist would crash when calling it on ath10k dual band
device set currently to use 2.4GHz. As freqlist worked, adapted txpowerlist
code to be the same as freqlist one. This fixes the issue.

Signed-off-by: Michal Hrusecky <michal.hrusecky@nic.cz>
---
 iwinfo_cli.c     | 4 ++--
 iwinfo_nl80211.c | 9 +++++++--
 2 files changed, 9 insertions(+), 4 deletions(-)

diff --git a/iwinfo_cli.c b/iwinfo_cli.c
index aab6ae6..7f95a26 100644
--- a/iwinfo_cli.c
+++ b/iwinfo_cli.c
@@ -611,7 +611,7 @@ static void print_scanlist(const struct iwinfo_ops *iw, const char *ifname)
 
 static void print_txpwrlist(const struct iwinfo_ops *iw, const char *ifname)
 {
-	int len, pwr, off, i;
+	int len = 0, pwr, off, i;
 	char buf[IWINFO_BUFSIZE];
 	struct iwinfo_txpwrlist_entry *e;
 
@@ -629,7 +629,7 @@ static void print_txpwrlist(const struct iwinfo_ops *iw, const char *ifname)
 
 	for (i = 0; i < len; i += sizeof(struct iwinfo_txpwrlist_entry))
 	{
-		e = (struct iwinfo_txpwrlist_entry *) &buf[i];
+		e = (struct iwinfo_txpwrlist_entry *) (buf + i);
 
 		printf("%s%3d dBm (%4d mW)\n",
 			(pwr == e->dbm) ? "*" : " ",
diff --git a/iwinfo_nl80211.c b/iwinfo_nl80211.c
index 63b70d5..ad02b32 100644
--- a/iwinfo_nl80211.c
+++ b/iwinfo_nl80211.c
@@ -1861,8 +1861,12 @@ static int nl80211_get_txpwrlist_cb(struct nl_msg *msg, void *arg)
 			nla_parse(freqs, NL80211_FREQUENCY_ATTR_MAX,
 			          nla_data(freq), nla_len(freq), freq_policy);
 
-			ch_cmp = nl80211_freq2channel(nla_get_u32(
-				freqs[NL80211_FREQUENCY_ATTR_FREQ]));
+			if (!freqs[NL80211_FREQUENCY_ATTR_FREQ] ||
+			    freqs[NL80211_FREQUENCY_ATTR_DISABLED])
+				continue;
+
+			ch_cmp = nla_get_u32(freqs[NL80211_FREQUENCY_ATTR_FREQ]);
+			ch_cmp = nl80211_freq2channel(ch_cmp);
 
 			if ((!ch_cur || (ch_cmp == ch_cur)) &&
 			    freqs[NL80211_FREQUENCY_ATTR_MAX_TX_POWER])
@@ -1884,6 +1888,7 @@ static int nl80211_get_txpwrlist(const char *ifname, char *buf, int *len)
 	int dbm_max = -1, dbm_cur, dbm_cnt;
 	struct nl80211_msg_conveyor *req;
 	struct iwinfo_txpwrlist_entry entry;
+	*len = 0;
 
 	if (nl80211_get_channel(ifname, &ch_cur))
 		ch_cur = 0;
-- 
2.10.1

